File: 2018/Python/Day01.py
import os

def part1(nums):
    return sum(nums)

def part2(nums):
    found = False
    current = 0
    freqs = {
        current: True
    }
    i = 0
    while(not found):
        current += nums[i % len(nums)]
        if current in freqs:
            found = True
        else:
            freqs[current] = True
        i += 1
    return current

def readFile(filename):
    path = os.path.abspath(__file__)
    scr_name = os.path.basename(__file__)
    with open(path.replace(scr_name, filename), 'r') as f:
        nums = f.read().splitlines()
    return [int(n) for n in nums]

if __name__ == "__main__":
    nums = readFile("Day01.txt")
    print(part1(nums))
    print(part2(nums))

File: 2018/Lua/Day01.lua
function part1(nums)
    local sum = 0
    for _, num in pairs(nums) do
        sum = sum + num
    end
    return sum
end

function part2(nums)
    local current = 0
    local seen = {current = true}
    local idx = 0
    while true do
        current = current + nums[(idx % #nums)+1]
        idx = idx + 1
        if seen[current] == nil then
            seen[current] = true
        else
            break
        end
    end
    return current
end

function main()
    local lines = read_file("Day01.txt")
    print(part1(lines))
    print(part2(lines))
end

-- HELPER FUNCTIONS
function file_exists(file)
    local f = io.open(file, "rb")
    if f then 
        f:close()
    end
    return f ~= nil
end

function read_file(file)
    if not file_exists(file) then 
        return {} 
    end
    lines = {}
    for line in io.lines(file) do
        lines[#lines+1] = tonumber(line)
    end
    return lines
end

main()